### PR TITLE
loops-nested_loops-hard-q7

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/loops/nested_loops/hard/q7/BoxMeNow.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package loops.nested_loops.hard.q7;
+
+/*
+Write a program called BoxMeNow to ask the use for n, and print the following pattern using nested loops based on their
+input. If the input is less than zero print “Invalid input!”
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/BoxMeNow.java:5]:
```suggestion
input. If the input is less than zero print “Invalid input!”.
```

File: src/loops/nested_loops/hard/q7/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void constructPatternCorrectly(int n) throws InvalidClauseException {
+        runWithInput(String.valueOf(n), clauseBuilder(n));
+    }
+
+    @Test
+    void identifiesInvalidInput() throws InvalidClauseException{
+        runWithInput("-49", new Clause[]{
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/MainTest.java:44]: See comments about boundary checking on the previous question.

File: src/loops/nested_loops/hard/q7/BoxMeNow.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+Sample Output for n =  4:
+Enter number of lines:
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/BoxMeNow.java:7]: This should probably be the side length of the box, as we don't want anyone to get confused and only print four at the top.

File: src/loops/nested_loops/hard/q7/BoxMeNow.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+4
+1111
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/BoxMeNow.java:9]: I think we should change the pattern to have a space between the ones so it actually will stay a box

Comment (Hedgemon4) [reply src/loops/nested_loops/hard/q7/BoxMeNow.java:9]: ie `1 1 1 1`

File: src/loops/nested_loops/hard/q7/BoxMeNow.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package loops.nested_loops.hard.q7;
+
+/*
+Write a program called BoxMeNow to ask the use for n, and print the following pattern using nested loops based on their
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/BoxMeNow.java:4]:
```suggestion
Write a program called BoxMeNow to ask the user for an n, and print the following box pattern using nested loops, with the n being the side length of the box (the number of 1's per side). 
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/nested_loops/hard/q7/BoxMeNow.java (hunk 3 above)
Comment (SamamthaH) [reply src/loops/nested_loops/hard/q7/BoxMeNow.java:7]: I'm confused. The number is the number on the sides of the box and the sample output shows it.

File: src/loops/nested_loops/hard/q7/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    public void runMain(){BoxMeNow.main(new String[0]);}
+
+    static Stream<Integer> inputProvider(){
+        return Stream.of(1, 0, 5, 23, 2);
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void constructPatternCorrectly(int n) throws InvalidClauseException {
+        runWithInput(String.valueOf(n), clauseBuilder(n));
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/MainTest.java:38]: This is missing an error message

File: src/loops/nested_loops/hard/q7/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    static Stream<Integer> invalidInputProvider(){return Stream.of(-1, -47, -300);}
+
+    @ParameterizedTest
+    @MethodSource("invalidInputProvider")
+    void identifiesInvalidInput(int in) throws InvalidClauseException{
+        runWithInput(String.valueOf(in), new Clause[]{
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/MainTest.java:46]: This is missing an error message

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/nested_loops/hard/q7/MainTest.java (hunk 6 above)
Comment (SamamthaH) [reply src/loops/nested_loops/hard/q7/MainTest.java:38]: I figure the standard wrong format would work here since the program is all format.

Comment (Hedgemon4) [reply src/loops/nested_loops/hard/q7/MainTest.java:38]: Same as the previous pull request, I think it is basically always better to specify an error message when possible.

Comment (Hedgemon4) [reply src/loops/nested_loops/hard/q7/MainTest.java:38]: It just allows us to target the feedback towards the things we test for, as opposed to if they just spell a word or something wrong.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/nested_loops/hard/q7/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    public void runMain(){BoxMeNow.main(new String[0]);}
+
+    static Stream<Integer> inputProvider(){
+        return Stream.of(1, 0, 5, 23, 2);
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void constructPatternCorrectly(int n) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the box correctly.";
```
Comment (Hedgemon4) [inline src/loops/nested_loops/hard/q7/MainTest.java:38]:
```suggestion
        TestOption.incorrectStructureErrorMessage = "Your program does not print the box pattern correctly.";
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**